### PR TITLE
Add sharedfx and targeting pack tests (#23045)

### PR DESCRIPTION
**Pull request by JunTaoLuo**
Addresses https://github.com/dotnet/aspnetcore/issues/19151 and https://github.com/dotnet/aspnetcore/issues/14832. Backport of https://github.com/dotnet/aspnetcore/pull/23045

### CODE REVIEW

**Review by dougbu — APPROVED** [review]
Please test the targeting pack part of this offline before merging, if you haven't already. Not building them right now.

And, to confirm: The hard-coded lists of expected assemblies isn't quite identical to what we're looking for in 'master', correct❔

**Comment by JunTaoLuo** [comment]
Yea I'll be tweaking a few of the details before merging.

File: src/Framework/test/SharedFxTests.cs
Comment (JunTaoLuo) [inline src/Framework/test/SharedFxTests.cs:39]: Hmm now I'm wondering why this test didn't fail in master without adding this.

Comment (JunTaoLuo) [reply src/Framework/test/SharedFxTests.cs:39]: Ah it's probably because it's testing the targeting pack that was built on Windows, given that we run the test on Helix. It should be benign then since we only ship the package built on Windows anyway.

**Comment by JunTaoLuo** [comment]
Tested locally to ensure the tests would pass when run. Made a few changes that warrant another look.

File: src/Framework/test/TargetingPackTests.cs
Comment (dougbu) [inline src/Framework/test/TargetingPackTests.cs:86]: This will blow up when extra assemblies are in the targeting pack. It's sufficient that `TargetingPackContainsListedAssemblies` fails in that case. Suggest using `TryGetValue(...)` and checking its return value.

**Review by dougbu — CHANGES_REQUESTED** [review]

File: src/Framework/test/SharedFxTests.cs
Comment (JunTaoLuo) [reply src/Framework/test/SharedFxTests.cs:39]: Good points. I'll address 1. As for 2, the reason is that we are running the tests on Helix and we use the test payload that was built on Windows even if we run the tests on non-Windows. @HaoK do we have infrastructure to build test payloads on non-Windows OS's. I don't remember having anything like that so maybe we should be running the tests on AzDO until we can do that?

File: src/Framework/test/SharedFxTests.cs
Comment (JunTaoLuo) [reply src/Framework/test/SharedFxTests.cs:35]: Yea it would probably work too.

File: src/Framework/test/SharedFxTests.cs
Comment (HaoK) [reply src/Framework/test/SharedFxTests.cs:39]: We don't have infrastructure to build on non windows yet, so maybe this test should have SkipTests explicitly set to false so it runs on azdo as well, since the behavior is different.

Comment (HaoK) [reply src/Framework/test/SharedFxTests.cs:39]: These seem like a good candidate for tests that we just never run on helix too, if they are testing installer/platform specific things?

Comment (JunTaoLuo) [reply src/Framework/test/SharedFxTests.cs:39]: I'll make sure the tests run when I merge this forward to master.

**Comment by JunTaoLuo** [comment]
:up::date:

**Review by dougbu — APPROVED** [review]

**Comment by JunTaoLuo** [comment]
@wtgodbe can go into release/3.1 right now? I just want to double check it wouldn't interfere with any of your 3.1 work.

**Comment by wtgodbe** [comment]
@JunTaoLuo please wait for next month's release, the build for 3.1.6 is already pretty much done

**Pull request merged**
